source/translated_table_model.ts

```diff
@@ -28,22 +45,158 @@ export class TranslatedTableModel extends TableModel {
+    }
+    if(this.references[row] >= 0) {
+      return this.model.get(this.references[row], column);
+    } else {
+      return this.model.get(row, column);
+    }
   }
 
   public connect(slot: (operations: Operation[]) => void):
       Kola.Listener<Operation[]> {
-    return null;
+    return this.dispatcher.listen(slot);
+  }
+
+  private handleOperations(newOperations: Operation[]): void {
+    this.beginTransaction();
+    for(const operation of newOperations) {
+      if(operation instanceof AddRowOperation) {
+        this.rowAdded(operation);
+      } else if (operation instanceof RemoveRowOperation) {
+        this.rowRemoved(operation);
+      } else if (operation instanceof UpdateValueOperation) {
+        this.updateRow(operation);
+      } else {
+        throw TypeError();
+      }
+    }
+    this.endTransaction();
+  }
+
+  private slideUp(start: number, end: number) {
+    for(let index = start; index < end; ++index) {
+      if(this.references[index + 1] >= 0) {
+        this.references[index] = this.references[index + 1];
+      } else {
+        this.references[index] = index + 1;
+      }
+    }
+  }
+
+  private slideDown(start: number, end: number) {
+    if(start === end) {
+      return;
+    }
+    if(end > start) {
+      throw RangeError();
+    }
+    for(let index = start; index > end; --index) {
+      if(this.references[index - 1] >= 0) {
+        this.references[index] = this.references[index - 1];
+      } else {
+        this.references[index] = index - 1;
+      }
+    }
+  }
+
+  private rowAdded(operation: AddRowOperation) {
+    this.beginTransaction();
+    this.references.push(undefined);
+    let referenceIndex = operation.index;
+    for(let index = 0; index < this.references.length; ++index) {
+      if(this.references[index] === operation.index) {
+        referenceIndex = index;
+      }
+    }
+    this.slideDown(this.references.length - 1, referenceIndex);
+    for(let index = 0; index < this.references.length; ++index) {
+      if(this.references[index] >= operation.index
+          && index !== referenceIndex) {
+        this.references[index] = this.references[index] + 1;
+      }
+    }
+    const row = new TranslatedTableModel(operation.row);
+    this.operations.push(new AddRowOperation(referenceIndex, row));
+    this.endTransaction();
+  }
+
+  private rowRemoved(operation: RemoveRowOperation) {
+    this.beginTransaction();
+    let referenceIndex = operation.index;
+    for(let index = 0; index < this.references.length; ++index) {
+      if(this.references[index] === operation.index) {
+        referenceIndex = index;
+      }
+    }
+    this.slideUp(referenceIndex, this.references.length);
+    this.references.pop();
+    for(let index = 0; index < this.references.length; ++index) {
+      if(this.references[index] > operation.index) {
+        this.references[index] = this.references[index] - 1;
+      }
+    }
+    const row = new TranslatedTableModel(operation.row);
```

Comment:
Don't wrap the row in a ```TranslatedTableModel```, in fact there's no reason why the row needs to be wrapped at all.